### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by shadowhand**
Rework document based on current interfaces. Add more detail regarding
design decisions and clarify intent to focus on server middleware.

### CODE REVIEW

**Comment by shadowhand** [comment]
This has been updated heavily to reflect recent decisions regarding naming and scope. Any feedback would be greatly appreciated.

cc @http-interop/http-middleware-contributors

File: README.md [hunk 1]
```diff
@@ -1,49 +1,56 @@
...
+
+[psr7]: http://www.php-fig.org/psr/psr-7/
 
 2. Why Bother?
 --------------
 
-The general concept of reusable middleware was popularized by [StackPHP][stackphp].
-Since the release of the HTTP Messages standard, a number of frameworks have
-adopted middleware that uses HTTP Message interfaces.
+The HTTP Messages specification contain any reference to HTTP Middleware.
```
Comment (weierophinney) [inline README.md]: `s/contain/do not contain/`.

File: README.md [hunk 2]
```diff
@@ -1,49 +1,56 @@
...
+
+The design pattern used by middleware has existed for many years as [pipeline][pipeline],
```
Comment (weierophinney) [inline README.md]: `s/[pipeline][pipeline]/[the pipeline pattern][pipeline]/`

File: README.md [hunk 3]
```diff
@@ -110,19 +119,17 @@ This may be resolved in the future with [functional interfaces][php-functional].
...
-* The middleware is defined as a [callable][php-callable] using a [closure][php-closure]
-  or a class with an `__invoke` method.
+* The middleware is defined with a specific interface with a method that takes
+  the request for processing.
 * The middleware is passed 2 arguments during invocation:
-  1. A `RequestInterface` implementation.
-  3. A `callable` that receives the request to dispatch next middleware.
-
+  1. An object that represents an HTTP request.
+  2. A delegate that receives the request to dispatch next middleware.
```
Comment (weierophinney) [inline README.md]: `s/dispatch next middleware/dispatch the next middleware in the pipeline/`.

File: README.md [hunk 4]
```diff
@@ -136,59 +143,133 @@ only the request being passed to the middleware.
...
+#### 4.2.2 Additional Projects Using Single Pass
+
+There are also significant projects that predate HTTP Messages using this approach.
+
+[StackPHP][stackphp] is based on [Symfony HttpKernel][httpkernel] and supports
+middleware with this signature:
 
 ```
-function handle(Request $request, callable $next): Response
+handle(Request $request, $type, $catch): Response
```
Comment (weierophinney) [inline README.md:163]: This contradicts line 119, where you indicated that the StackPHP "style" is `fn(request, frame): response`. I'm thinking that the narrative at line 119 should likely not mention Stack at all.

File: README.md [hunk 5]
```diff
@@ -136,59 +143,133 @@ only the request being passed to the middleware.
...
+#### Why is a server request required?
+
+To make it clear that the middleware can only be used in a synchronous, server
+side context.
+
+While not all middleware will need to use the additional methods defined by the
+server request interface, external requests are typically handled asynchronously
+and would need to return a promise of a response. This is outside the scope of
+this proposal and would complicate matters for any future proposal that is focused
+on client side request processing.
```
Comment (weierophinney) [inline README.md]: I think this paragraph needs some expansion:
- You gloss over "external requests". Maybe expand to say, "For client-side middleware that makes a request to an external resource, you will typically want to react asychronously...", and then detail _why_ async is typical in this situation.
- Why is a "promise of a response" required under async? (_I_ know, but not everyone is familiar with async processing.)
- What is a "promise"? (_I_ know, but will every reader coming to the spec? Link to the [Promises A+ spec](https://promisesaplus.com/)

**Review by weierophinney — CHANGES_REQUESTED** [review]
I've proposed a few wording changes, and then noted some areas that are either slightly contradictory, or which need expansion.

**Review by mindplay-dk — APPROVED** [review]

**Comment by mindplay-dk** [comment]
What I mean by my approval here is, "approved with minor comments", just the minor corrections suggested by @weierophinney.

File: README.md (hunk 4 above)
Comment (shadowhand) [reply README.md:163]: Stack implements has multiple parameters, but none of them are `$response`. As such, I think it still constitutes a widely used example of single-pass.

**Pull request merged**